File: aoc-2024/day14/Day14.java
import java.util.Scanner;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.stream.IntStream;
import java.util.List;
import java.util.Objects;

class Pair {
    private final int x;
    private final int y;

    Pair(int x, int y) {
        this.x = x;
        this.y = y;
    }

    int x() {
        return this.x;
    }

    int y() {
        return this.y;
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        }
        if (obj instanceof Pair other) {
            return other.x == this.x && other.y == this.y;
        }
        return false;
    }

    @Override
    public int hashCode() {
        return Objects.hash(this.x, this.y);
    }
}

class Robot {
    private final Pair pos;
    private final Pair vel;

    public Robot(Pair pos, Pair vel) {
        this.pos = pos;
        this.vel = vel;
    }

    public Pair position() {
        return this.pos;
    }

    public Pair velocity() {
        return this.vel;
    }
}

class Simulator {
    private final int width;
    private final int height;
    private final int count;
    private List<Robot> robots;

    public Simulator(int width, int height) {
        this.width = width;
        this.height = height;
        this.count = 0;
        this.robots = List.of();
    }

    public Simulator(int width, int height, int count, List<Robot> robots) {
        this.width = width;
        this.height = height;
        this.count = count;
        this.robots = robots;
    }

    private int calculatePos(int p, int v, int s, int m) {
        int r = (p + (v * s)) % m;
        if (r < 0) {
            r = m + r;
        }
        return r;
    }

    public List<Pair> calculateFor(int duration) {
        return this.robots.stream().parallel()
            .map(r -> new Pair(calculatePos(r.position().x(), r.velocity().x(), duration, this.width),
                                calculatePos(r.position().y(), r.velocity().y(), duration, this.height)))
            .toList();
    }

    public void displayFloor() {
        ConcurrentHashMap<Pair, AtomicInteger> map = new ConcurrentHashMap<>(this.width * this.height);

        this.robots.stream().parallel()
            .forEach(r -> {
                map.computeIfAbsent(r.position(), f -> new AtomicInteger(0)).incrementAndGet();
            });

        for (int y = 0; y < this.height; y += 1) {
            for (int x = 0; x < this.width; x += 1) {
                AtomicInteger c = map.getOrDefault(new Pair(x, y), new AtomicInteger(0));
                if (c.get() == 0) {
                    System.out.print(".");
                } else {
                    System.out.print(c);
                }
            }
            System.out.println("");
        }
        System.out.printf("Count: %d\n", this.count);
    }

    public Simulator nextState() {
        return new Simulator(this.width, this.height, this.count + 1, this.robots.stream().parallel()
            .map(r -> {
                int nx = this.calculatePos(r.position().x(), r.velocity().x(), 1, this.width);
                int ny = this.calculatePos(r.position().y(), r.velocity().y(), 1, this.height);
                return new Robot(new Pair(nx, ny), r.velocity());
            })
            .toList()
        );
    }
}

class Day14 {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        List<Robot> initRobots = sc.useDelimiter("\n")
            .tokens()
            .parallel()
            .map(line -> parseInput(line))
            .toList();
        sc.close();
        int width = 101;
        int height = 103;
        int duration = 10000;
        Simulator ns = new Simulator(width, height, 0, initRobots);

        System.out.printf("Part One: %d \n", calculatePartOne(ns.calculateFor(duration), width, height));

        runSimulator(ns, duration, width, height);
    }

    private static Robot parseInput(String input) { // p=x,y v=x,y
        String[] ss = input.trim().split(" ");
        String[] ps = ss[0].split("=")[1].split(",");
        String[] vs = ss[1].split("=")[1].split(",");
        return new Robot(
            new Pair(Integer.parseInt(ps[0]), Integer.parseInt(ps[1])),
            new Pair(Integer.parseInt(vs[0]), Integer.parseInt(vs[1]))
        );
    }

    private static int calculatePartOne(List<Pair> finalPos, int width, int height) {
        int q1 = 0; // x = 0 ~ (width / 2 - 1), y = 0 ~ (height / 2 - 1)
        int q2 = 0; // x = (width / 2 + 1) ~ (width - 1), y = 0 ~ (height / 2 - 1)
        int q3 = 0; // x = 0 ~ (width / 2 - 1), y = (height / 2 + 1) ~ (height - 1)
        int q4 = 0; // x = (width / 2 + 1) ~ (width - 1), y = (height / 2 + 1) ~ (height - 1)

        List<Integer> quadrants = finalPos.stream()
            .map(p -> belongsTo(p, width, height))
            .toList();

        q1 = (int)quadrants.stream().parallel().filter(q -> q == 1).count();
        q2 = (int)quadrants.stream().parallel().filter(q -> q == 2).count();
        q3 = (int)quadrants.stream().parallel().filter(q -> q == 3).count();
        q4 = (int)quadrants.stream().parallel().filter(q -> q == 4).count();
        return q1 * q2 * q3 * q4;
    }

    private static int belongsTo(Pair pos, int width, int height) {
        if ((pos.x() >= 0 && pos.x() <= (width / 2 - 1)) &&
            (pos.y() >= 0 && pos.y() <= (height / 2 - 1))) {
            return 1;
        } else if ((pos.x() >= (width / 2 + 1) && (pos.x() <= width - 1)) &&
                    (pos.y() >= 0 && pos.y() <= (height / 2 - 1))) {
            return 2;
        } else if ((pos.x() >= 0 && pos.x() <= (width / 2 - 1)) &&
                    (pos.y() >= (height / 2 + 1) && pos.y() <= height - 1)) {
            return 3;
        } else if ((pos.x() >= (width / 2 + 1) && (pos.x() <= width - 1)) &&
                    (pos.y() >= (height / 2 + 1) && (pos.y() <= height - 1))) {
            return 4;
        }
        return -1;
    }

    private static void runSimulator(Simulator ini, int maxRound, int width, int height) {
        ini.displayFloor();
        System.out.println("=".repeat(50));
        for (int i = 1; i <= maxRound; i += 1) {
            ini = ini.nextState();
            if ((i - height) % width == 0) {
                System.out.println(i);
                ini.displayFloor();
                System.out.println("=".repeat(50));
            }
        }
    }
}

File: aoc-2024/day14/day14.go
package main

import (
    "bufio"
    "fmt"
    "os"
    "strings"
)

type Pair struct {
	x int
	y int
}

func main() {
    var rawInputs []string
    var inputs [][2]Pair
    var err error
    if rawInputs, err = readInput(); err == nil {
        if inputs, err = parseInput(rawInputs); err != nil {
            fmt.Println(err)
            return
        }
    } else {
        fmt.Println(err)
        return
    }

	width := 101
	height := 103
	factor := calculatePartOne(inputs, 100, width, height)
    fmt.Printf("Part One: %d\n", factor)

	runSimulation(inputs, 10000, width, height)
}

// Part One

func calculatePartOne(robots [][2]Pair, time, width, height int) int {
	finalPos := calculatePos(robots, time, width, height)
	quadrants := calculateQuadrant(finalPos, width, height)
	return quadrants[0] * quadrants[1] * quadrants[2] * quadrants[3]
}

func calculatePos(robots [][2]Pair, time, width, height int) []Pair {
	var pos []Pair = []Pair{}
	for _, r := range(robots) {
		nx := (r[0].x + (r[1].x * time)) % width
		ny := (r[0].y + (r[1].y * time)) % height
		if (nx < 0) {
			nx += width
		}
		if (ny < 0) {
			ny += height
		}
		pos = append(pos, Pair{x: nx, y: ny})
	}
	return pos
}

func calculateQuadrant(pos []Pair, width, height int) [4]int {
	var quadrants [4]int = [4]int{0, 0, 0, 0}
	for _, p := range(pos) {
		if ((p.x >= 0 && p.x <= (width / 2 - 1)) &&
            (p.y >= 0 && p.y <= (height / 2 - 1))) {
			quadrants[0] += 1
        } else if ((p.x >= (width / 2 + 1) && (p.x <= width - 1)) &&
                    (p.y >= 0 && p.y <= (height / 2 - 1))) {
            quadrants[1] += 1
        } else if ((p.x >= 0 && p.x <= (width / 2 - 1)) &&
                    (p.y >= (height / 2 + 1) && p.y <= height - 1)) {
            quadrants[2] += 1
        } else if ((p.x >= (width / 2 + 1) && (p.x <= width - 1)) &&
                    (p.y >= (height / 2 + 1) && (p.y <= height - 1))) {
            quadrants[3] += 1
        }
	}
	return quadrants
}

// Part Two

func runSimulation(robots [][2]Pair, time, width, height int) {
	displayFloor(robots, width, height)
	fmt.Println(strings.Repeat("=", 50))
	for i := range(time) {
		robots = nextState(robots, width, height)
		if ((i + 1 - height) % width == 0) {
			displayFloor(robots, width, height)
			fmt.Printf("Count: %d\n", i + 1)
			fmt.Println(strings.Repeat("=", 50))
		}
	}
}

func nextState(robots [][2]Pair, width, height int) [][2]Pair {
	for i, r := range(robots) {
		r[0].x = (r[0].x + r[1].x) % width
		r[0].y = (r[0].y + r[1].y) % height
		if (r[0].x < 0) {
			r[0].x += width
		}
		if (r[0].y < 0) {
			r[0].y += height
		}
		robots[i] = r
	}
	return robots
}

func displayFloor(robots [][2]Pair, width, height int) {
	posMap := make(map[Pair]int, width * height)
	for _, r := range(robots) {
		pos := r[0]
		_, ok := posMap[pos]
		if ok {
			posMap[pos] += 1
		} else {
			posMap[pos] = 1
		}
	}
	for y := range(height) {
		for x:= range(width) {
			count, ok := posMap[Pair{x: x, y: y}]
			if (ok && count > 0) {
				fmt.Print(count)
			} else {
				fmt.Print(".")
			}
		}
		fmt.Println("")
	}
}

func parseInput(rawInputs []string) ([][2]Pair, error) {
    var parseResult [][2]Pair = [][2]Pair{}
    //var parseResult [][4]int = [][4]int{}
    var err error
    for _, line := range(rawInputs) {
        var in [2]Pair = [2]Pair{Pair{}, Pair{}}
        //var in [4]int = [4]int{}
        var temp []string = strings.Split(line, " ")

		for i, pv := range(temp) {
			var nums []string = strings.Split(strings.Split(pv, "=")[1], ",")
			_, err = fmt.Sscanf(nums[0], "%d", &in[i].x)
			if (err != nil) {
				return [][2]Pair{}, err
				//return [][4]int{}, err
			}

			_, err = fmt.Sscanf(nums[1], "%d", &in[i].y)
			if (err != nil) {
				//return [][4]int{}, err
				return [][2]Pair{}, err
			}
		}
        parseResult = append(parseResult, in)
    }
    return parseResult, nil
}

func readInput() ([]string, error) {
    var input []string
    var err error
    var scanner bufio.Scanner = *bufio.NewScanner(os.Stdin)

    for {
        scanner.Scan()
        var line string = scanner.Text()

        if (len(line) == 0) {
            break
        }

        input = append(input, line)
    }

    err = scanner.Err()
    if (err != nil) {
        fmt.Println("Error occured when trying to read user input!")
        return input, err
    }

    return input, nil
}
